source/crud/crud.rst

```diff
@@ -1927,8 +1928,67 @@ The ``update`` family of operations require that the update document parameter M
 The ``replace`` family of operations require that the replacement document parameter MUST NOT begin with an atomic modifier. In practice, this means that introspection needs to happen on that document to enforce this. However, it is enough to only check the first element in the document. If it does not begin with a ``$`` sign but an element later on does, the server will throw an error.
 
 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Aggregation Pipelines with Write Stages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section discusses special considerations for aggregation pipelines that
+contain write stages (e.g. ``$out``, ``$merge``).
+
+
+Returning a cursor on the output collection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As noted in the documentation for the ``aggregate`` helper earlier in this
+document, ``$out`` and ``$merge`` are special pipeline stages that cause no
+results to be returned from the server. As such, drivers MAY setup a cursor to
+be executed upon iteration against the output collection and return that instead
+of an iterable that would otherwise have no results.
+
+Drivers that do so for ``$merge`` MAY remind users that such a cursor may return
+more documents than were written by the aggregation (e.g. documents that existed
+in the collection prior to ``$merge`` being executed).
+
+
+Read preferences and server selection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Historically, only primaries could execute an aggregation pipeline with ``$out``
+or ``$merge``. As of ``featureCompatibilityVersion`` 4.4, secondaries can now
+execute such pipelines; however, since drivers do not track
+``featureCompatibilityVersion``, executing an aggregation pipeline with ``$out``
+or ``$merge`` on a secondary will require a 5.0+ server (i.e. wire version >=
+13).
+
+Drivers MUST guarantee the following:
+
+- If an explicit (i.e. per-operation) read preference is specified for an
+  aggregation with a write stage, drivers MUST attempt to use it. If that would
+  result in a pre-5.0, secondary server being selected, drivers MUST instead
+  select a server using a primary read preference.
+- If no explicit read preference is specified but a default read preference is
+  available to inherit (e.g. from the Collection), drivers MUST attempt to use
+  it. If that would result in a pre-5.0, secondary server being selected,
+  drivers MUST instead select a server using a primary read preference.
+
+Since the server version may not be known prior to server selection, this
+specification is not prescriptive about how drivers should select a server for
+an ``aggregate`` command including a write stage. For example, drivers MAY use a
+custom server selector to consider server/wire version when matching a read
+preference and fall back to selecting a primary. Alternatively, drivers MAY
```

Comment:
Does this MAY imply that it's ok for a driver to effectively double the server selection timeout, or do you have to subtract the time for the first server selection before falling back to the second?  If so, the spec should indicate that explicitly.  Additionally, the spec should indicate what should happen if the initial server selection times out.  I imagine at that point the driver has to error, even if the second server selection would have succeeded (assuming we would have fallen back had the first selection succeeded).

Reply:
I have no qualms about duplicating the server selection timeout in the interest of minimizing driver complexity here. If that does prove to be a concern, I expect CSOT would also address it in due time; however, I'm not expecting this to be a problem given the small audience for this feature. AFAIK, the only current demand for this feature is for internal users of the Go driver with Charts.

I do think that stopping if the first server selection attempt times out is reasonable, although it is a notable difference from the single attempt approach (with a custom selector). For drivers that do need to make two attempts, catching and ignoring the first attempt's error would also introduce undesirable complexity.

So, proposed changes would be:

 * Explicitly note that the two-attempt approach may result in the total selection time exceeding `serverSelectionTimeoutMS`
 * Require drivers using the two-attempt approach to forgo the second attempt if the first resulted in a server selection timeout. It's _possible_ we could find a primary on the second attempt but another timeout may be more likely (especially for a permissive read preference like `secondaryPreferred`).

Does anyone else have thoughts on the matter? If not, I'll follow up on Monday with spec changes.

Reply:
>  however, I'm not expecting this to be a problem given the small audience for this feature

I'm not so sure about that.  Given the way we're exposing this feature, we're effectively opting applications into it.  Any application with a non-default read preference defined for client/database/collection will hit this code path, even if they are still on 4.4-and-below server and always end up with the fallback-to-primary scenario.

Reply:
> Given the way we're exposing this feature, we're effectively opting applications into it.

I agree. "small audience" was a poor choice of words on my part. The feature itself is frustrating to me, as it seemingly violates so many existing rules we have around server selection.

I wasn't around for scoping, but I would have much preferred to never have `$out`/`$merge` inherit a read preference and only allow drivers to start accepting a per-operation read preference for them. But that would leave drivers w/o per-operation read preference options no way to support this at all.

> Any application with a non-default read preference defined for client/database/collection will hit this code path, even if they are still on 4.4-and-below server and always end up with the fallback-to-primary scenario.

I don't think there's _any_ performance regression for drivers implementing a single attempt with custom selection logic (i.e. considering read preference and wire version at once). For drivers that decide to make a second attempt as needed, let's enumerate the possible code paths by topology type:

 * Single topology: it's only possible to select a single server, so a second attempt never makes sense. It's possible that the application has a direct connection to a secondary (or some non-writable server). Perhaps the spec should clarify that no second attempt should ever be made for a single topology.
 * Replica set topology: a second attempt only happens if a pre-5.0 secondary is selected. If a primary (any version) or 5.0+ secondary is selected, that server is always usable and no second attempt is needed.
 * Sharded or load-balanced topology: the first and only selection attempt is going to return a server capable of receiving a write. Based on the wire version of mongos (either the direct node or the backing mongos), the driver will decide to forward the original read preference or substitute `mode=primary`.

If my understanding is correct, I think this only affects a subset of applications that (a) use a read preference that can potentially select a secondary _and_ (b) have pre-5.0 secondaries in the cluster _and_ (c) performance aggregations with `$out`/`$merge`.

Reply:
To recap, proposed changes to address the two-attempt approach:

 * Drivers MUST NOT make a second attempt if the first attempt encounters a server selection timeout.
 * Drivers MUST NOT make second attempts for single, sharded, or load-balanced topologies (for reasons cited above, there is no reason to do so).
 * Explicitly discuss how the two-attempt approach can result in a performance regression (as discussed in the end of [last comment](https://github.com/mongodb/specifications/pull/1062#discussion_r706937170)).

Reply:
Is it an option to make implementation with a single server selection attempt a MUST?

Reply:
I would only go as far as SHOULD. Making it a requirement would require creating an entirely new server selection API in libmongoc just to support the single-attempt approach in PHPLIB. I'm not convinced that this feature warrants the engineering effort.

Reply:
@kevinAlbs what do you think?

Reply:
It would be nice to have a level between SHOULD and MUST - ideally, we'd require drivers to not do two rounds of server selection as long as the technical complexity that's involved doesn't justify it.

For PHPLIB specifically, I don't think the complexity of exposing custom server selection in libmongoc and PHPC is worth it. I don't believe that a lot of users would run into the double server selection in PHPLIB, but we can always use PHP's error reporting mechanism (`trigger_error`) to warn users of a potential performance instance that they can then handle as they see fit.

Of course, we can always use MUST and allow deviations from the spec, so basically "drivers MUST NOT perform two rounds of server selection, but MAY choose to do so if the complexity involved doesn't justify it" (probably the worst wording, but I hope it gets the point across so others can improve it)

Reply:
> It would be nice to have a level between SHOULD and MUST - ideally, we'd require drivers to not do two rounds of server selection as long as the technical complexity that's involved doesn't justify it.

I do think that SHOULD fits the bill for this. It _is_ actually quite strict on its own, as drivers are not expected to deviate from a SHOULD without good reason (technical complexity in this case). To that end, I'd consider something like:

> drivers SHOULD use the single-attempt approach with custom server selection logic, but MAY use two attempts if...

Reply:
SHOULD with a following MAY seems accurate to me; I also agree that overhauling server selection in libmongoc is not worth it for this feature. Go should be able to implement the single-attempt approach with a custom server selector.

Reply:
Updated these paragraphs:

 * Drivers SHOULD use the single attempt approach with custom logic
 * Drivers MAY use the two attempt approach if necessary

For the two attempt approach:

 * Drivers MUST NOT make a second attempt if the first results in a timeout error. Instead, they raise the original error.
 * Drivers MUST NOT make a second attempt for Single, Sharded, or LoadBalanced topologies. There's no reason to do so, as the first server would either be writable (most cases) or the only option (i.e. Single topology with a direct connection to a pre-5.0 secondary).

Reply:
> Making it a requirement would require creating an entirely new server selection API in libmongoc just to support the single-attempt approach in PHPLIB.

I agree with that assessment. I considered whether it would be sufficient to add the [custom server selector client level option from the Server Selection specification](https://github.com/mongodb/specifications/blob/c48d7f416e51dce8fd4c8d05bdb05790508ca3bb/source/server-selection/server-selection.rst#serverselector). But the custom callback would not know if the current operation is an agg with `$out`/`$merge`. So the API addition would be specific to this feature.

I am not opposed new API to libmongoc if it is needed for correctness. But if the caveat is doubling server selection timeout in rare scenarios, my vote goes for less implementation complexity. CSOT will likely require a deadline in server selection, so we could remove this caveat later.